File: workflow/05single_script/helpers/h5ad_contract_utils.R
`%||%` <- function(x, y) {
  if (is.null(x) || length(x) == 0L) y else x
}

h5ad_contract_repo_template_path <- function() {
  workflow_root <- Sys.getenv("WORKFLOW_ROOT", unset = "")
  if (nzchar(workflow_root)) {
    return(file.path(workflow_root, "..", "metadata", "h5ad_export_contract.tsv.template"))
  }
  file.path(getwd(), "metadata", "h5ad_export_contract.tsv.template")
}

load_h5ad_contract <- function(contract_path = Sys.getenv("H5AD_CONTRACT_FILE", "metadata/h5ad_export_contract.tsv"), modality = c("scrna", "spatial")) {
  modality <- match.arg(modality)
  if (!file.exists(contract_path)) {
    contract_path <- h5ad_contract_repo_template_path()
  }
  if (!file.exists(contract_path)) {
    stop(sprintf("missing H5AD contract: %s", contract_path), call. = FALSE)
  }
  contract <- read.delim(contract_path, sep = "\t", comment.char = "#", stringsAsFactors = FALSE, check.names = FALSE)
  validate_h5ad_contract_schema(contract)
  contract[contract$modality %in% c(modality, "both"), , drop = FALSE]
}

validate_h5ad_contract_schema <- function(contract) {
  required_cols <- c("field_path", "field_type", "required", "dtype", "gate", "modality", "semantics", "notes")
  missing_cols <- setdiff(required_cols, colnames(contract))
  if (length(missing_cols) > 0L) {
    stop(sprintf("H5AD contract missing columns: %s", paste(missing_cols, collapse = ", ")), call. = FALSE)
  }

  allowed_type <- c("scalar", "matrix", "dict", "dataframe")
  allowed_required <- c("yes", "no", "conditional")
  allowed_gate <- c("primary", "exploratory", "conditional")
  allowed_modality <- c("scrna", "spatial", "both")
  bad_type <- unique(contract$field_type[!contract$field_type %in% allowed_type])
  bad_required <- unique(contract$required[!contract$required %in% allowed_required])
  bad_gate <- unique(contract$gate[!contract$gate %in% allowed_gate])
  bad_modality <- unique(contract$modality[!contract$modality %in% allowed_modality])
  if (length(bad_type) > 0L) stop(sprintf("invalid field_type: %s", paste(bad_type, collapse = ", ")), call. = FALSE)
  if (length(bad_required) > 0L) stop(sprintf("invalid required: %s", paste(bad_required, collapse = ", ")), call. = FALSE)
  if (length(bad_gate) > 0L) stop(sprintf("invalid gate: %s", paste(bad_gate, collapse = ", ")), call. = FALSE)
  if (length(bad_modality) > 0L) stop(sprintf("invalid modality: %s", paste(bad_modality, collapse = ", ")), call. = FALSE)

  core <- c("obs.sample_id", "obs.condition", "var.gene_id", "layers.counts")
  missing_core <- setdiff(core, contract$field_path)
  if (length(missing_core) > 0L) {
    stop(sprintf("H5AD contract removed core fields: %s", paste(missing_core, collapse = ", ")), call. = FALSE)
  }
  invisible(TRUE)
}

is_mock_h5ad_object <- function(obj) {
  is.list(obj) && any(c("obs", "var", "obsm", "layers", "uns", "spatial") %in% names(obj))
}

contract_location_field <- function(field_path) {
  parts <- strsplit(field_path, ".", fixed = TRUE)[[1]]
  list(location = parts[[1]], field = paste(parts[-1], collapse = "."))
}

mock_contract_value <- function(obj, field_path) {
  parsed <- contract_location_field(field_path)
  location <- parsed$location
  field <- parsed$field
  container <- obj[[location]]
  if (is.null(container)) {
    return(list(found = FALSE, value = NULL))
  }
  if (is.data.frame(container)) {
    if (field %in% colnames(container)) return(list(found = TRUE, value = container[[field]]))
    return(list(found = FALSE, value = NULL))
  }
  if (is.list(container)) {
    parts <- strsplit(field, ".", fixed = TRUE)[[1]]
    cursor <- container
    for (part in parts) {
      if (!is.list(cursor) || is.null(cursor[[part]])) {
        return(list(found = FALSE, value = NULL))
      }
      cursor <- cursor[[part]]
    }
    return(list(found = TRUE, value = cursor))
  }
  list(found = FALSE, value = NULL)
}

seurat_contract_value <- function(seurat_obj, field_path) {
  parsed <- contract_location_field(field_path)
  location <- parsed$location
  field <- parsed$field
  value <- NULL
  found <- FALSE

  if (location == "obs" && !is.null(seurat_obj@meta.data) && field %in% colnames(seurat_obj@meta.data)) {
    found <- TRUE
    value <- seurat_obj@meta.data[[field]]
  } else if (location == "var") {
    assay_name <- if (exists("DefaultAssay", mode = "function")) DefaultAssay(seurat_obj) else names(seurat_obj@assays)[[1]]
    meta_features <- seurat_obj@assays[[assay_name]]@meta.features
    if (!is.null(meta_features) && field %in% colnames(meta_features)) {
      found <- TRUE
      value <- meta_features[[field]]
    }
  } else if (location == "obsm") {
    reductions <- if (exists("Reductions", mode = "function")) Reductions(seurat_obj) else names(seurat_obj@reductions)
    reduction_name <- sub("^X_", "", field)
    if (field %in% reductions || reduction_name %in% reductions) {
      found <- TRUE
      value <- matrix(0, nrow = ncol(seurat_obj), ncol = 1)
    }
  } else if (location == "layers") {
    assay_name <- if (exists("DefaultAssay", mode = "function")) DefaultAssay(seurat_obj) else names(seurat_obj@assays)[[1]]
    assay <- seurat_obj@assays[[assay_name]]
    if (field %in% c("counts", "data", "scale.data", "logcounts", "SCT")) {
      found <- TRUE
      value <- tryCatch({
        if (field %in% slotNames(assay)) slot(assay, field) else matrix(0, nrow = 1, ncol = 1)
      }, error = function(e) matrix(0, nrow = 1, ncol = 1))
    }
  } else if (location == "uns" && !is.null(seurat_obj@misc[[field]])) {
    found <- TRUE
    value <- seurat_obj@misc[[field]]
  } else if (location == "spatial" && length(seurat_obj@images) > 0L) {
    found <- TRUE
    value <- seurat_obj@images
  }

  list(found = found, value = value)
}

detect_h5ad_contract_dtype <- function(value) {
  if (is.null(value)) return("missing")
  if (is.data.frame(value)) return("dataframe")
  if (is.list(value) && !is.data.frame(value)) return("dict")
  if (is.matrix(value) || inherits(value, "Matrix")) {
    if (is.integer(value)) return("int64")
    if (is.numeric(value)) return("float32")
    if (is.character(value)) return("str")
    return(class(value)[[1]])
  }
  if (is.integer(value)) return("int64")
  if (is.numeric(value)) return("float32")
  if (is.character(value) || is.factor(value)) return("str")
  if (is.logical(value)) return("bool")
  class(value)[[1]]
}

dtype_compatible_h5ad_contract <- function(actual, expected) {
  expected <- expected %||% ""
  if (!nzchar(expected) || expected == "-") return(TRUE)
  if (actual == expected) return(TRUE)
  if (expected %in% c("float32", "float64") && actual %in% c("float32", "float64", "int64")) return(TRUE)
  if (expected %in% c("int64", "int32") && actual %in% c("int64", "int32")) return(TRUE)
  if (expected == "str" && actual %in% c("str", "factor")) return(TRUE)
  FALSE
}

contract_violation_row <- function(field, severity, reason, expected_dtype = "", actual_dtype = "") {
  data.frame(
    field = field,
    severity = severity,
    reason = reason,
    expected_dtype = expected_dtype,
    actual_dtype = actual_dtype,
    stringsAsFactors = FALSE
  )
}

check_h5ad_contract_seurat <- function(seurat_obj, contract, fail_on = c("error", "warn", "none")) {
  fail_on <- match.arg(fail_on)
  validate_h5ad_contract_schema(contract)
  violations <- list()
  is_mock <- is_mock_h5ad_object(seurat_obj)

  for (i in seq_len(nrow(contract))) {
    row <- contract[i, , drop = FALSE]
    field_path <- row$field_path[[1]]
    required <- row$required[[1]]
    gate <- row$gate[[1]]
    severity <- if (gate == "primary") "error" else "warn"
    value_info <- if (is_mock) mock_contract_value(seurat_obj, field_path) else seurat_contract_value(seurat_obj, field_path)

    if (!isTRUE(value_info$found)) {
      if (required == "yes") {
        parsed <- contract_location_field(field_path)
        violations[[length(violations) + 1L]] <- contract_violation_row(
          field_path,
          severity,
          sprintf("required=yes but field not found in %s", parsed$location),
          row$dtype[[1]],
          "missing"
        )
      }
      next
    }

    actual_dtype <- detect_h5ad_contract_dtype(value_info$value)
    if (!dtype_compatible_h5ad_contract(actual_dtype, row$dtype[[1]])) {
      violations[[length(violations) + 1L]] <- contract_violation_row(
        field_path,
        severity,
        sprintf("expected dtype=%s got %s", row$dtype[[1]], actual_dtype),
        row$dtype[[1]],
        actual_dtype
      )
    }
  }

  vdf <- if (length(violations) > 0L) do.call(rbind, violations) else data.frame(
    field = character(), severity = character(), reason = character(),
    expected_dtype = character(), actual_dtype = character(), stringsAsFactors = FALSE
  )
  passed <- !any(vdf$severity == "error")
  result <- list(
    passed = passed,
    violations = vdf,
    summary = data.frame(
      passed = passed,
      violation_n = nrow(vdf),
      error_n = sum(vdf$severity == "error"),
      warn_n = sum(vdf$severity == "warn"),
      stringsAsFactors = FALSE
    ),
    contract = contract
  )

  if (!passed && fail_on == "error") {
    stop(sprintf(
      "H5AD contract violations: %s",
      paste(vdf$field[vdf$severity == "error"], collapse = ", ")
    ), call. = FALSE)
  }
  if (nrow(vdf) > 0L && fail_on == "warn") {
    warning(sprintf("H5AD contract violations: %d field(s)", nrow(vdf)), call. = FALSE)
  }
  result
}

contract_violations_to_manifest <- function(check_result) {
  violations <- check_result$violations %||% data.frame()
  list(
    contract_passed = isTRUE(check_result$passed),
    contract_violations_count = nrow(violations),
    contract_violations_summary = lapply(seq_len(nrow(violations)), function(i) as.list(violations[i, , drop = FALSE]))
  )
}

File: workflow/04python/helpers/h5ad_contract_utils.py
from __future__ import annotations

import csv
import os
import warnings
from pathlib import Path
from typing import Any

REQUIRED_COLUMNS = [
    "field_path",
    "field_type",
    "required",
    "dtype",
    "gate",
    "modality",
    "semantics",
    "notes",
]
ALLOWED_FIELD_TYPES = {"scalar", "matrix", "dict", "dataframe"}
ALLOWED_REQUIRED = {"yes", "no", "conditional"}
ALLOWED_GATES = {"primary", "exploratory", "conditional"}
ALLOWED_MODALITIES = {"scrna", "spatial", "both"}
CORE_FIELDS = {"obs.sample_id", "obs.condition", "var.gene_id", "layers.counts"}


def repo_template_path() -> Path:
    return Path(__file__).resolve().parents[3] / "metadata" / "h5ad_export_contract.tsv.template"


def load_h5ad_contract(contract_path: str | Path | None = None, modality: str = "scrna") -> list[dict[str, str]]:
    if modality not in {"scrna", "spatial"}:
        raise ValueError(f"invalid modality: {modality}")
    path = Path(contract_path or os.environ.get("H5AD_CONTRACT_FILE", "metadata/h5ad_export_contract.tsv"))
    if not path.exists():
        path = repo_template_path()
    if not path.exists():
        raise FileNotFoundError(f"missing H5AD contract: {path}")

    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader((line for line in handle if not line.startswith("#")), delimiter="\t")
        rows = [dict(row) for row in reader]
    validate_h5ad_contract_schema(rows)
    return [row for row in rows if row["modality"] in {modality, "both"}]


def validate_h5ad_contract_schema(rows: list[dict[str, str]]) -> None:
    if not rows:
        raise ValueError("empty H5AD contract")
    missing_columns = [column for column in REQUIRED_COLUMNS if column not in rows[0]]
    if missing_columns:
        raise ValueError(f"H5AD contract missing columns: {', '.join(missing_columns)}")

    fields = {row["field_path"] for row in rows}
    missing_core = sorted(CORE_FIELDS - fields)
    if missing_core:
        raise ValueError(f"H5AD contract removed core fields: {', '.join(missing_core)}")

    for row in rows:
        if row["field_type"] not in ALLOWED_FIELD_TYPES:
            raise ValueError(f"invalid field_type: {row['field_type']}")
        if row["required"] not in ALLOWED_REQUIRED:
            raise ValueError(f"invalid required: {row['required']}")
        if row["gate"] not in ALLOWED_GATES:
            raise ValueError(f"invalid gate: {row['gate']}")
        if row["modality"] not in ALLOWED_MODALITIES:
            raise ValueError(f"invalid modality: {row['modality']}")


def _split_field(field_path: str) -> tuple[str, str]:
    location, _, field = field_path.partition(".")
    return location, field


def _nested_get(mapping: Any, dotted: str) -> tuple[bool, Any]:
    cursor = mapping
    for part in dotted.split("."):
        if isinstance(cursor, dict) and part in cursor:
            cursor = cursor[part]
        else:
            return False, None
    return True, cursor


def _value_from_mapping(mapping: dict[str, Any], field_path: str) -> tuple[bool, Any]:
    location, field = _split_field(field_path)
    if location not in mapping:
        return False, None
    container = mapping[location]
    if hasattr(container, "columns") and field in container.columns:
        return True, container[field]
    if isinstance(container, dict):
        return _nested_get(container, field)
    return False, None


def _value_from_anndata(adata: Any, field_path: str) -> tuple[bool, Any]:
    location, field = _split_field(field_path)
    if location == "obs":
        if field in getattr(adata, "obs").columns:
            return True, adata.obs[field]
    elif location == "var":
        if field in getattr(adata, "var").columns:
            return True, adata.var[field]
    elif location == "obsm":
        if field in getattr(adata, "obsm").keys():
            return True, adata.obsm[field]
    elif location == "layers":
        if field in getattr(adata, "layers").keys():
            return True, adata.layers[field]
    elif location == "uns":
        return _nested_get(getattr(adata, "uns"), field)
    elif location == "spatial":
        spatial = getattr(adata, "uns", {}).get("spatial", {})
        return _nested_get(spatial, field)
    return False, None


def detect_dtype(value: Any) -> str:
    if value is None:
        return "missing"
    if hasattr(value, "dtypes") and hasattr(value, "columns"):
        return "dataframe"
    if isinstance(value, dict):
        return "dict"
    dtype = getattr(value, "dtype", None)
    if dtype is not None:
        dtype_text = str(dtype)
        if "int" in dtype_text:
            return "int64"
        if "float" in dtype_text:
            return "float32"
        if "bool" in dtype_text:
            return "bool"
        if "object" in dtype_text or "str" in dtype_text:
            return "str"
    if isinstance(value, (list, tuple)) and value:
        if all(isinstance(item, (list, tuple)) for item in value):
            flat = [cell for row in value for cell in row]
            if flat and all(isinstance(item, int) and not isinstance(item, bool) for item in flat):
                return "int64"
            if flat and all(isinstance(item, (int, float)) and not isinstance(item, bool) for item in flat):
                return "float32"
        if all(isinstance(item, int) and not isinstance(item, bool) for item in value):
            return "int64"
        if all(isinstance(item, (int, float)) and not isinstance(item, bool) for item in value):
            return "float32"
        if all(isinstance(item, str) for item in value):
            return "str"
    if isinstance(value, str):
        return "str"
    if isinstance(value, bool):
        return "bool"
    if isinstance(value, int):
        return "int64"
    if isinstance(value, float):
        return "float32"
    return type(value).__name__


def dtype_compatible(actual: str, expected: str) -> bool:
    if not expected or expected == "-":
        return True
    if actual == expected:
        return True
    if expected in {"float32", "float64"} and actual in {"float32", "float64", "int64", "int32"}:
        return True
    if expected in {"int64", "int32"} and actual in {"int64", "int32"}:
        return True
    return expected == "str" and actual == "str"


def check_h5ad_contract_anndata(adata: Any, contract: list[dict[str, str]], fail_on: str = "warn") -> dict[str, Any]:
    if fail_on not in {"error", "warn", "none"}:
        raise ValueError(f"invalid fail_on: {fail_on}")
    validate_h5ad_contract_schema(contract)
    is_mapping = isinstance(adata, dict)
    violations: list[dict[str, str]] = []

    for row in contract:
        field_path = row["field_path"]
        found, value = _value_from_mapping(adata, field_path) if is_mapping else _value_from_anndata(adata, field_path)
        severity = "error" if row["gate"] == "primary" else "warn"
        if not found:
            if row["required"] == "yes":
                location, _ = _split_field(field_path)
                violations.append({
                    "field": field_path,
                    "severity": severity,
                    "reason": f"required=yes but field not found in {location}",
                    "expected_dtype": row["dtype"],
                    "actual_dtype": "missing",
                })
            continue

        actual = detect_dtype(value)
        if not dtype_compatible(actual, row["dtype"]):
            violations.append({
                "field": field_path,
                "severity": severity,
                "reason": f"expected dtype={row['dtype']} got {actual}",
                "expected_dtype": row["dtype"],
                "actual_dtype": actual,
            })

    primary_failed = any(item["severity"] == "error" for item in violations)
    if primary_failed and fail_on == "error":
        fields = ", ".join(item["field"] for item in violations if item["severity"] == "error")
        raise ValueError(f"H5AD contract violations: {fields}")
    if violations and fail_on == "warn":
        warnings.warn(f"H5AD contract violations: {len(violations)} field(s)", stacklevel=2)
    return {
        "passed": not primary_failed,
        "violations": violations,
        "summary": {
            "violation_n": len(violations),
            "error_n": sum(item["severity"] == "error" for item in violations),
            "warn_n": sum(item["severity"] == "warn" for item in violations),
        },
    }


def contract_violations_to_manifest(check_result: dict[str, Any]) -> dict[str, Any]:
    violations = check_result.get("violations", [])
    return {
        "contract_passed": bool(check_result.get("passed")),
        "contract_violations_count": len(violations),
        "contract_violations_summary": violations,
    }
